### PR TITLE
Fetch and sync track updates when a webhook comes in (V2)

### PR DESCRIPTION
**Pull request by kntsoriano**
## Description
This is a reimplementation of fetching and syncing track updates. This was done because the first PR didn't take into account of the following:

1. Track update fetches need to happen on all webservers. The previous PR only does it on the webserver that receives the webhook.
2. Syncing should only happen when all webservers have fetched the track update. The previous PR has no way to guarantee this.

## Solution
Taking the above into account, here is a summary of what this implements:

1. When a webhook comes in, a `TrackUpdate` model is created. The `TrackUpdate` model is only created when it is a push to master.
2. The webservers poll for a fetchable `TrackUpdate`. A `TrackUpdate` is fetchable if it has no associated `TrackUpdateFetch` records that matches the webserver's host name.
3. When the webserver fetches a `TrackUpdate`, a `TrackUpdateFetch` record is created recording the webserver's host name. After fetching, the same `TrackUpdateFetch` record is marked with the time it has been fetched.
4. The processor polls for a syncable `TrackUpdate`. A `TrackUpdate` is syncable if the `TrackUpdate` has fetched `TrackUpdateFetches` equal to the number of webservers. This is currently set as `ENV[NUM_WEBSERVERS]` on the processor.
5. After the processor syncs a `TrackUpdate`, the `TrackUpdate` record is marked with the time it has been synced.

## Discussion
1. Since we set `ENV[NUM_WEBSERVERS]` on the processor, we should find a way to update this when we add a new webserver. I think we can do this later.

### CODE REVIEW

**Comment by ccare** [comment]
I love the way you communicate on your PRs @kntsoriano, so clear. Thank you! 

Regarding, the number of webservers, I wonder if we should do this as a yaml file, rather than an env var. This isn't really a strong opinion, but based on two things:

1) By using a config file we can ensure it's set for every rails engine we spawn (webserver, rake task, sidekiq processor). I get that it's probably not needed everywhere, but it feels prudent to ensure that.

2) We then have only one approach to managing external config on exercism

Don't get me wrong, I love ENV vars, and they totally make sense, so this can be done if you feel strongly. What do you think?

File: app/services/git/syncs_updated_tracks.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  end
+
+  def self.sync
+    new.sync
+  end
+
+  def run
+    loop do
+      sync
+      puts "Sleeping"
```
Comment (ccare) [inline app/services/git/syncs_updated_tracks.rb]: I need to get the rails logger working properly for rake tasks. It's a good idea to stick with PUTS for now, but I'm hoping to have that sorted today.

File: app/jobs/sync_track_update_job.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  after_perform do |job|
+    track_update = TrackUpdate.find(job.arguments.first)
+
+    track_update.update!(synced_at: Time.current)
+  end
+
+  def perform(track_update_id)
+    track_update = TrackUpdate.find(track_update_id)
+
+    Git::SyncsTracks.sync([track_update.track])
```
Comment (ccare) [inline app/jobs/sync_track_update_job.rb]: I wonder if we can just remove the `SyncsTracks` class now that we're not looping through everything? What do you think. There's obviously that hacky bit of code about `hello-world` to address, but that can go inside `SyncsTrack` I think.

File: app/services/git/syncs_updated_tracks.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  private
+  attr_reader :track_update
+
+  def find_track_update
+    @track_update = TrackUpdate.
+      joins(:track_update_fetches).
+      where(synced_at: nil).
+      where.not(track_update_fetches: { completed_at: nil }).
+      group("track_update_fetches.id").
+      having("count(track_update_id) = #{ENV["NUM_WEBSERVERS"]}").
```
Comment (ccare) [inline app/services/git/syncs_updated_tracks.rb]: Nice use of `having` there, I still find active record queries hard to do. As I mentioned in the top-level comment. I wonder if we should externalise this in a YAML file so it can be set on a per-env basis. That's the pattern we've been following throughout this project, but happy to be overruled if the env var has advantages. See https://github.com/exercism/prototype/blob/master/app/models/git/github_profile.rb#L4 and https://github.com/exercism/prototype/blob/master/config/auth.yml for an example. It might be we want to refactor some of these configs and have one for managing the git integration stuff, maybe...

File: app/services/git/fetches_tracks.rb [hunk 4]
```diff
@@ -1,25 +1,26 @@
...
-        sleep 1.second
-      end
-      sleep QUIET_PERIOD
+  def initialize(tracks)
+    @tracks = tracks
+  end
+
+  def fetch
+    fetch_problem_specs
+    fetch_website_content
```
Comment (ccare) [inline app/services/git/fetches_tracks.rb]: We need to address this two fetches, because they're not tracks, but are specialisations of Git::RepoBase and therefore need to be fetched and triggered by a webhook - sorry for the scope creep!

File: app/services/git/fetches_updated_tracks.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  private
+
+  attr_reader :track_update, :track_update_fetch
+
+  def find_track_update
+   @track_update = TrackUpdate.
+     includes(:track_update_fetches).
+     where(synced_at: nil).
+     find_by(track_update_fetches: { track_update_id: nil }) ||
```
Comment (ccare) [inline app/services/git/fetches_updated_tracks.rb]: I find your code very readable @kntsoriano. This method did confuse me a bit though, wonder if the `||` here could be made a little clearer, I didn't spot it at first.

File: app/services/git/syncs_tracks.rb [hunk 6]
```diff
@@ -1,41 +1,24 @@
 class Git::SyncsTracks
```
Comment (ccare) [inline app/services/git/syncs_tracks.rb]: I think this class is almost redundant now. Same for `FetchesTracks` possibly, and the `Git::StateDb`. Your solution is much better.

File: app/jobs/fetch_track_update_job.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
+class FetchTrackUpdateJob < ApplicationJob
+  attr_accessor :track_update_fetch_id
+
+  before_perform do |job|
+    track_update = TrackUpdate.find(job.arguments.first)
+
+    track_update_fetch = TrackUpdateFetch.create!(
+      track_update: track_update,
+      host: Socket.gethostname
```
Comment (ccare) [inline app/jobs/fetch_track_update_job.rb]: We need to think about Socket.gethostname in an ec2 context, but let's defer that until we've sorted everything else out. There are a couple of options here which would be fun to discuss separately.

File: app/services/git/fetches_tracks.rb [hunk 8]
```diff
@@ -1,25 +1,26 @@
...
+    tracks.each do |track|
+      fetch_track(track)
+      sleep 1.second
```
Comment (ccare) [inline app/services/git/fetches_tracks.rb]: This my code not yours, but I should probably have extracted the backoff into a constant.

File: app/controllers/webhooks/track_updates_controller.rb [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  def create
+    TrackUpdate.create(track: track) if pushed_to_master?
+
+    render json: {}, status: 200
+  end
+
+  private
+
+  def pushed_to_master?
+    params[:ref].eql? MASTER_REF
```
Comment (ccare) [inline app/controllers/webhooks/track_updates_controller.rb]: 👍 I suspect I would have missed this subtlety

File: app/services/git/fetches_tracks.rb (hunk 4 above)
Comment (ccare) [inline app/services/git/fetches_tracks.rb]: Maybe `TrackUpdateFetches` should be generalised to `GitUpdateFetches` ?

File: app/jobs/sync_track_update_job.rb (hunk 2 above)
Comment (kntsoriano) [reply app/jobs/sync_track_update_job.rb]: In the previous PR, you mentioned a nightly task that pulls tracks just in case the webhooks don't work. I think having the `SyncsTracks` service helps there since we can only call `SyncsTracks.sync(Track.all)`.

File: app/services/git/syncs_updated_tracks.rb (hunk 3 above)
Comment (kntsoriano) [reply app/services/git/syncs_updated_tracks.rb]: Sure, I don't know any advantages of having an ENV var over a YAML file. I'll be glad to change this to use the YAML file.

File: app/services/git/fetches_tracks.rb (hunk 4 above)
Comment (kntsoriano) [reply app/services/git/fetches_tracks.rb]: I think we can stick with this code now and just generalize to `GitUpdateFetches` :)

File: app/services/git/fetches_tracks.rb (hunk 8 above)
Comment (kntsoriano) [reply app/services/git/fetches_tracks.rb]: No problem! I'll extract that to a constant.

File: app/services/git/fetches_updated_tracks.rb (hunk 5 above)
Comment (kntsoriano) [reply app/services/git/fetches_updated_tracks.rb]: Sure, I'll come up with a way.

File: app/services/git/syncs_tracks.rb (hunk 6 above)
Comment (kntsoriano) [reply app/services/git/syncs_tracks.rb]: I think we should still keep this in order to have an easy way to sync all tracks, just as I have described for `FetchesTracks`. What do you think?

**Comment by kntsoriano** [comment]
I have no opinions of using an ENV var over a YAML file. Since YAML files are what we use right now, I'll update my code to use it. Thanks for that :)

File: app/jobs/fetch_track_update_job.rb (hunk 7 above)
Comment (kntsoriano) [reply app/jobs/fetch_track_update_job.rb]: I think we can talk about this now. I've addressed your comments and I've got it working locally. This is the last bit we need to work on before we can deploy.

File: app/services/git/fetches_tracks.rb (hunk 4 above)
Comment (ccare) [reply app/services/git/fetches_tracks.rb]: I don't think so - what if just the website content changes? That should trigger a webhook which triggers a refresh of that content.

File: app/jobs/sync_track_update_job.rb (hunk 2 above)
Comment (ccare) [reply app/jobs/sync_track_update_job.rb]: Good point. I see

File: app/jobs/fetch_track_update_job.rb (hunk 7 above)
Comment (ccare) [reply app/jobs/fetch_track_update_job.rb]: so there's no harm using the hostname, but it'll change on reboot (unless we pin internal addresses). Alternatively, we could lookup the public IP from the EC2 api (static and straightforward), or we could get chef to write a file somewhere that configures a server slug.

Comment (ccare) [reply app/jobs/fetch_track_update_job.rb]: Would it matter if the hostname changed on reboot? I'm assuming so, but maybe not.

Comment (kntsoriano) [reply app/jobs/fetch_track_update_job.rb]: I think it matters because we need to know whether the host has fetched the server or not. Does the public IP change only when we use new instance? It could also be a problem when the machine's IP changes while the instance doesn't.

File: app/services/git/fetches_tracks.rb (hunk 4 above)
Comment (kntsoriano) [reply app/services/git/fetches_tracks.rb]: Good point. I'd need to write some code to account for the problem specs and website copy repos changing.

File: app/jobs/fetch_track_update_job.rb (hunk 7 above)
Comment (ccare) [reply app/jobs/fetch_track_update_job.rb]: We do have fairly static IPs, but lets decouple this from the network identity, because as you hint, it's problematic. I'll adjust chef to write a slug into `/etc/exercism/server_identity` so if you just read from that file (and chomp) then we're good I think.

**Comment by kntsoriano** [comment]
Hello, I figured that renaming tables and dropping foreign keys would be too complicated to continue on this branch. Please refer to the new pull request I opened: https://github.com/exercism/prototype/pull/16
